src/lib/components/Widget.js

```diff
@@ -0,0 +1,203 @@
+      this.setState({
+        isDarkMode: browserDarkMode ? true : false,
+      });
+    }
+  }
+
+  capitalizeFirstLetter(stringToCapitalize) {
+    return (
+      stringToCapitalize.charAt(0).toUpperCase() + stringToCapitalize.slice(1)
+    );
+  }
+
+  extractFromRuleContent(term, text) {
+    var start = text.substring(text.search(`${term}:`));
+    var value = start.substring(term.length + 1, start.search("\\n"));
+    return value.trim();
+  }
+
+  async setRules() {
+    const pullRequests = await this.fetchPullRequests();
+
+    var filesToRetrieve = [];
+    var additionalFileDetails = [];
+
+    for (let pr of pullRequests) {
+      for (let file of pr.files.nodes) {
+        if (
+          !filesToRetrieve.includes(file.path) &&
+          (file.path.substring(file.path.length - 3) === ".md" ||
+            file.path.substring(file.path.length - 9) === ".markdown") &&
+          file.path.substring(0, 6) === "rules/"
+        ) {
+          filesToRetrieve = [...filesToRetrieve, file.path];
+
+          additionalFileDetails = [
+            ...additionalFileDetails,
+            {
+              author: pr.author.login,
+              timestamp: new Date(pr.mergedAt),
+            },
+          ];
+        }
+      }
+    }
+
+    var retrievedFileContents = await this.fetchFileContents(filesToRetrieve);
+    for (let [i, file] of retrievedFileContents.entries()) {
+      if (file != null) {
+        var title = this.extractFromRuleContent("title", file);
+        var uri = this.extractFromRuleContent("uri", file);
+
+        this.setState({
+          isLoaded: true,
+          rules: [
+            ...this.state.rules,
+            {
+              id: this.state.rules.length + 1,
+              uri: uri,
+              path: file.path,
+              title: title,
+              author: additionalFileDetails[i].author,
+              timestamp: additionalFileDetails[i].timestamp,
+            },
+          ],
+        });
+      }
+    }
+  }
+
+  async fetchPullRequests() {
+    const pullRequests = await requestPullRequests(
+      this.state.numberOfRules,
+      this.state.author,
+      this.state.token
+    );
+    pullRequests.sort((a, b) => new Date(b.mergedAt) - new Date(a.mergedAt));
+    return pullRequests;
+  }
+
+  async fetchFileContents(filesToRetrieve) {
+    var fileContents = await requestMultipleFileContents(
+      filesToRetrieve,
+      this.state.numberOfRules,
+      this.state.token
+    );
+
+    if (
+      fileContents.filter((x) => x !== null).length < this.state.numberOfRules
+    ) {
+      var counter = 0;
+
+      while (
+        fileContents.filter((x) => x !== null).length < this.state.numberOfRules
+      ) {
+        var extraFile = await requestSingleFileContents(
+          filesToRetrieve[counter + this.state.numberOfRules],
+          this.state.token
+        );
+        fileContents = [...fileContents, extraFile];
+        counter++;
+      }
+    }
```

Comment:
What is the point of this if statement, I think you can remove the if and just call while